### ISSUE TITLE
proposal: cmd/go: support for Git LFS via new VCS qualifier

### ISSUE THREAD

**Issue by jpap** [golang/go#47308]
## Background

* `cmd/go` works well with Git LFS as-is only when it is installed on the client machine.  
* When Git LFS is not installed, `go get` will download the repository but not smudge, so any files stored in Git LFS are left as text-file pointer files.  This can break builds when such files are required for the build (e.g. syso files).
* The Google-hosted `proxy.golang.org` uses `cmd/go` to download package contents.  That service does not have Git LFS installed, and so every module retrieved via the proxy does not have the Git LFS-hosted content included.  This can break builds.  It can also result in `sumdb.golang.org` mismatches when a user tries to verify a package downloaded with Git LFS to one downloaded by GOPROXY (without Git LFS).
* For further detailed background information, see #47241.

## Proposal

1. A new VCS named `git-lfs` that prescribes the use of Git LFS when fetching a Git repo.

2. New `cmd/go` support for the new VCS: using a built-in Git LFS client implementation, or one that relies on Git LFS installed on the client machine.  For the latter:
    * If not installed, `cmd/go` should inform the user that Git LFS is required before erroring out.  i.e. `cmd/go` should never leave un-smudged Git LFS pointer files in a fetched package.
    * Have Git LFS installed on `proxy.golang.org`.

3. Package (module) owners opt-in to mandatory Git LFS package retrieval using the following methods:
    * Specifying `git-lfs` as the `vcs` argument in the `content` attribute of the [go-import meta tag](https://golang.org/ref/mod#vcs-find).
    * As a VCS qualifier to the package import path [see footnote 1]:
        * In the `module` directive in a `go.mod` file, e.g. `module go.jpap.org/example.git-lfs`
        * In the import comment in a Go source file, e.g. `package example // import "go.jpap.org/example.git-lfs"`

        The VCS qualifier here guides how `cmd/go` fetches the package (module).  It forms the suffix to a module path as defined in the [GOPROXY protocol](https://golang.org/ref/mod#goproxy-protocol) (see `$module`).  It should, however, be ignored when validating the package (module) path against an `import` statement in a Go source file.  The example above should match a Go source file with an `import "go.jpap.org/example"` statement.

4. Package (module) users can explicitly fetch a Git LFS package by specifying the proposed VCS on the command line, e.g. `go get go.jpap.org/example.git-lfs`.  This is helpful when there is no `go-import` HTTPS server available.

5. Where Git LFS is installed, explicitly disable smudging on `cmd/go` fetches of packages (modules) not opted into the `git-lfs` VCS qualifier as outlined in Item 3.

## Footnotes

1. When a module is retrieved by `go/cmd` in direct mode, `go-import` yields no results, and the retrieved `go.mod` file has the `git-lfs` VCS qualifier, `cmd/go` should automatically "upgrade" to a Git LFS fetch.

    Similarly, if a GOPATH package was requested by `go get`, `go-import` yields no results, and the retrieved package has an import comment with a `git-lfs` VCS qualifier, it too should be "upgraded" to a Git LFS fetch.

    A Git LFS fetch "upgrade" here means performing a Git LFS smudge on the downloaded files to replace refs with their actual content.  If `cmd/go` relies on Git LFS being installed to do this, it should error out if unavailable.

    This procedure allows for package (module) users to automatically allow `import` (`require`) statements without change, for example, when upgrading a new package (module) dependency version that uses `git-lfs`.

**Comment by ianlancetaylor** [comment]
CC @bcmills @jayconrod

**Comment by bcmills** [comment]
> in the import comment in a package, e.g. `package example // import "go.jpap.org/example.git-lfs"`.

FWIW, import comments are ignored in module mode. If this behavior falls out naturally from the implementation that's fine, but I think the only places we should explicitly support the extension are in the module path:
```
$ go get go.jpap.org/example.git-lfs/cmd/example
```
and in the `go-import` metadata:
```html
<meta name="go-import" content="go.jpap.org/example git-lfs https://github.com/jpap/example.git">
```
.

**Comment by bcmills** [comment]
> When fetching a module (package), `cmd/go` should "upgrade" to Git LFS when it detects that the package has opted into Git LFS. In that case, the module path (package import path) should be treated as if the user had appended the `.git-lfs` suffix.

I don't think we should do this part — it's too much of a special case — but I _do_ agree that we need to provide a smoother migration path for existing projects that don't control the `go-import` metadata for their import path, such as users with `github.com` import paths.

(I'm actively working on a proposal for a more general solution to the “renamed module path” use-case, so stay tuned!)

**Comment by bcmills** [comment]
All in all, though, I agree that adding a `git-lfs` VCS type seems like the cleanest path forward, and explicitly disabling LFS for the regular `git` VCS type is the least likely to break existing checksums.

**Comment by jpap** [comment]
> > in the import comment in a package, e.g. `package example // import "go.jpap.org/example.git-lfs"`.
> 
> FWIW, import comments are ignored in module mode. If this behavior falls out naturally from the implementation that's fine, but I think the only places we should explicitly support the extension are in the module path:

That was specified in support of GOPATH packages.  I understand that GOPATH is deprecated, but is it going to be removed and become strictly unsupported?

> > ["upgrade" stuff] 
>
> I don't think we should do this part — it's too much of a special case...

I don't feel strongly about that: I've dropped Item 5 of the proposal, and appended an "error out" item in its place.

> ... but I _do_ agree that we need to provide a smoother migration path for existing projects that don't control the `go-import` metadata for their import path, such as users with `github.com` import paths.
>
> (I'm actively working on a proposal for a more general solution to the “renamed module path” use-case, so stay tuned!)

You mentioned this back in #47241 -- are you able to outline some of what you had in mind for those like me who are curious?  I hope this is related to #39536!

**Comment by bcmills** [comment]
> That was specified in support of GOPATH packages. I understand that GOPATH is deprecated, but is it going to be removed and become strictly unsupported?

We're not doing new feature work on it, and don't plan to. I think we should implement `git-lfs` support in module mode, and if that happens to also make it work in `GOPATH` mode all the better — but we shouldn't do extra work (or define extra semantics) just to support the new protocol in the old mode.

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://golang.org/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by bcmills** [comment]
I think we should accept this. Specifically:
* `git-lfs` is treated as a new, independent VCS, meaning “Git with LFS definitely enabled, and fail if it isn't supported”
* `git` is treated as “Git with LFS explicitly disabled” (as `proxy.golang.org` does implicitly today by virtue of not having LFS installed)

I don't think we should try to detect when a repo served as `git` is configured for LFS — those repos may already exist and happen to work today and we shouldn't break them.

That does leave the hard-coded hosting providers like `github.com` in a bit of an awkward state, in that users can't seamlessly migrate them from `git` to `git-lfs` without changing the module path. But I think we can address that separately with follow-up proposals (#26134 is related, and #26904 may be sufficient once the `.git-lfs` suffix is recognized).

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://golang.org/s/proposal-status#likely-accept)**.
— rsc for the proposal review group

**Comment by rsc** [comment]
No change in consensus, so **[accepted](https://golang.org/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— rsc for the proposal review group

**Comment by sbinet** [comment]
it seems this proposal died out.
it would be very useful for my latest package ([daktilo](https://codeberg.org/sbinet/daktilo)) that wraps a rather large static archive (~150Mb).

has there been some PoC I could *try* to build upon ?

**Comment by apparentlymart** [comment]
I must admit I don't remember all of the feature release dates well enough to know for sure, but I wonder if this proposal was originally discussed before creating the precedent of using the `go` directive as a way to opt-in to new features, and if so whether that might be a less fiddly way to get this functionality.

Specifically, I mean:

- If `go.mod` selects a version greater than or equal to the Go version that introduced explicit Git LFS support, _and_ the repository has metadata that implies it's expecting to use Git LFS, then the module is assumed to require Git LFS and will fail to install if the needed client isn't available.

   As previously suggested, the systems where `proxy.golang.org` does its work would have Git LFS installed, so that they can support modules that match this situation.
- If `go.mod` selects a sufficiently-new version and there _isn't_ any Git LFS metadata in the repository then the toolchain won't try to run Git LFS and so it doesn't need to be installed.
- If `go.mod` selects an older version of Go that predates this change then the behavior matches that of the selected version of Go, and in particular the `proxy.golang.org` system would not run Git LFS in this case.

Changing the `go` directive in a `go.mod` requires publishing a new version of a module, and so this can't cause new treatment for any already-published module and thus existing modules will continue to match their original checksums.

I think this means that a sufficiently-new version of Go would handle Git LFS in a reasonable way by default, without needing to use any unusual patterns to force a different VCS implementation, and the `go.mod` settings can cause a sufficiently-new version of the Go toolchain to be automatically installed if it wasn't already.

All of this assumes that the main problem that motivated having a separate VCS was to avoid changing the meaning of already-published modules. If there were other reasons I'm not aware of then this alternative proposal might not work.

---

**Edit** on Nov 11 2025:

Thanks to discussion in another issue I found https://github.com/golang/go/issues/30369 which proposed using the `go` directive to select between different ways of deciding what is included in a module and thus how its checksum is calculated.

Although that was motivated by some different concerns, I think the reasons why it was retracted could also justify rejecting what I described above.

**Comment by sbinet** [comment]
I naïvely thought one would first need to implement LFS support for `cmd/go` before `proxy.golang.org`, but perhaps you're right that one should tackle this the other way around.

I just tried the workarounds listed in https://github.com/golang/go/issues/47241 and was met with inconsistent checksums:

```sh
$> mkdir use-daktilo
$> go mod init use-daktilo
$> GOPROXY=direct go get -v codeberg.org/sbinet/daktilo@6bbdc0a
get "codeberg.org/sbinet/daktilo": found meta tag vcs.metaImport{Prefix:"codeberg.org/sbinet/daktilo", VCS:"git", RepoRoot:"https://codeberg.org/sbinet/daktilo.git", SubDir:""} at //codeberg.org/sbinet/daktilo?go-get=1
go: downloading codeberg.org/sbinet/daktilo v0.3.2-0.20251024113904-6bbdc0a1c787
go: codeberg.org/sbinet/daktilo@6bbdc0a: codeberg.org/sbinet/daktilo@v0.3.2-0.20251024113904-6bbdc0a1c787: verifying module: checksum mismatch
	downloaded: h1:rUN1rSbZvMcUEFSFeMz0gIBC4BeGP/qjeiyUs7Ds05Y=
	sum.golang.org: h1:t1c3NZmPmraPlccobpXBHivjot5X6U70h/7NpyHX/7Q=

SECURITY ERROR
This download does NOT match the one reported by the checksum server.
The bits may have been replaced on the origin server, or an attacker may
have intercepted the download attempt.

For more information, see 'go help module-auth'.
```

it was only when I disabled sumdb that I could compile the package:
```
$> GOPROXY=direct GOSUMDB=off go get -v codeberg.org/sbinet/daktilo@6bbdc0a
get "codeberg.org/sbinet/daktilo": found meta tag vcs.metaImport{Prefix:"codeberg.org/sbinet/daktilo", VCS:"git", RepoRoot:"https://codeberg.org/sbinet/daktilo.git", SubDir:""} at //codeberg.org/sbinet/daktilo?go-get=1
go: downloading codeberg.org/sbinet/daktilo v0.3.2-0.20251024113904-6bbdc0a1c787
$> go build
$> echo $?
0
```

Obviously, disabling sumdb isn't advisable.

So, yes, LFS support probably needs to be gated by some Go version (presumably `1.27`) to not change the meaning/content of already-published modules (and their checksum).

Should we open a new issue or proposal to plan that work ?
(or regroup and devise a Go-specific way to bundle large binaries to be compiled w/ Go code ?)

**Comment by JAicewizard** [comment]
I have a semi working implementation: https://github.com/JAicewizard/go/tree/lfs

Using `GOSUMDB=off GOPROXY=direct {$GOROOT}/bin/go get -x gitlab.com/amfiremage/lfstest.git-lfs@latest` I get a local copy, including all LFS objects in `{$HOME}/go/pkg/mod/gitlab.com/amfiremage/lfstest.git-lfs@v0.0.0-20251228183327-0933000b7e5b`.

Sadly, it is not fully working yet, notably, it does not actually build...

I have no idea why, the error message is `{$HOME}/go/pkg/mod/gitlab.com/amfiremage/lfstest.git-lfs@v0.0.0-20251228183327-0933000b7e5b/test.go:1:1: expected 'package', found version`. Surprisingly, the listed file is in fact the correct go file, not some unsmudged file like the error indicates. This is likely because go build does not actually use this file(?)

- [x] A new VCS named git-lfs that prescribes the use of Git LFS when fetching a Git repo. (Technically missing, its called Git but I think Git LFS would be better, this is just a small TODO for me)
- [ ] New cmd/go support for the new VCS: using a built-in Git LFS client implementation, or one that relies on Git LFS installed on the client machine. For the latter:
  - [x] If not installed, cmd/go should inform the user that Git LFS is required before erroring out. i.e. cmd/go should never leave un-smudged Git LFS ref files in a fetched package. (Technically implemented, as `git lfs` is executed explicitly. Should have nicer upfront checks/message)
  - [ ] Have Git LFS installed on proxy.golang.org. (Not my job)

- [ ] Module (package) owners opt-in to mandatory Git LFS package retrieval using the following:
  - [x] Specifying git-lfs as the vcs argument in the content attribute of the [go-import meta tag](https://golang.org/ref/mod#vcs-find). (Untested, but the code that handles this is updated)
  - [ ] Using the new VCS qualifier,
        in the module directive in a go.mod file, e.g. module go.jpap.org/example.git-lfs.
    -  Here the .git-lfs suffix forms part of the module path when used with the [GOPROXY protocol](https://golang.org/ref/mod#goproxy-protocol) and the local module caches ($GOPATH/pkg/mod).
    -  in the import comment in a package, e.g. `package example // import "go.jpap.org/example.git-lfs"`.

    In both cases above, the .git-lfs suffix should be ignored when matching the module path (package import path) against an import statement in a Go source file. With reference to the examples above, the Go source having import statement `import "go.jpap.org/example"` should match even though it does not specify the VCS qualifier.

- [x] Module (package) users can explicitly qualify a module path (package import path) using the proposed VCS so that Git LFS is used to access the repository. This is helpful when there is no go-import HTTP server available to resolve the repository URL. (Note that the wording here is slightly wrong, it should be when there is no go-import HTTP server *that knows about the git-lfs qualifier*.)

- [x] Where Git LFS is installed, explicitly disable smudging on cmd/go fetches of modules (packages) not opted into the git-lfs VCS qualifier as outlined in Item 3. This prevents the checksum mismatch problem outlined in https://github.com/golang/go/issues/47241.

- [ ] Irrespective of whether Git LFS is installed, when fetching a module (package) cmd/go should error out when it detects that a package has opted into Git LFS and the VCS qualifier is not git-lfs. This forces module (package) users to also opt-in to using Git LFS, and prevents broken builds due to files not being smudged appropriately when Git LFS is not installed. (I don't fully understand this one, per definition when a module opts in it will have a lfs qualifier. That is how it opts in, no?)

As you can see, this is mostly done, only needing some small fixes and a nice error message. The bigger issue is builds not working, which is a shame it doesn't automatically just work, as I would expect.

**Comment by JAicewizard** [comment]
Update, seems like the build issue was just caching! Lots of wasted time, but this should all mostly work now!

**Comment by JAicewizard** [comment]
> With reference to the examples above, the Go source having import statement import "go.jpap.org/example" should match even though it does not specify the VCS qualifier.

I just realised this is tricky. What should happen when we have the following mod?:

```
...
require (
    gitlab.com/amfiremage/lfstest.git v0.0.1 // indirect
    gitlab.com/amfiremage/lfstest.git-lfs v0.0.1 // indirect
    gitlab.com/amfiremage/lfstest.hg v0.0.1 // indirect
)
```
There is no guarantee they resolve to the same module, in fact they are guaranteed to have different `go.mod` not the hg repo could have nothing to do with the git repos, besides sharing the name and version.

I think this is a legal `go.mod`, and should be handled correctly (not just guessing). (Besides that it might be impossible to have git and git-lfs like this.)

The code is also heavily geared towards import path and module path always including the vcs qualifier. You need to specify it in the module declaration etc etc. Making this work feels out of scope, and should be a separate proposal and or implementation. I do not feel comfortable doing that. I propose that this is left out for now, and this can be added later when these edge-cases are considered more cleanly.

Besides this, I will leave my branch as an RFC for now, please check it out, I will put it in gerit when I have time to do so!

**Comment by seankhliao** [comment]
https://github.com/golang/go/issues/47308#issuecomment-1122828551 is what we accepted.

Specifically, that means `.git`, `.git-lfs` must be part of the declared module path and import path when used.

```
require gitlab.com/amfiremage/lfstest.git v0.0.1
```
corresponds to a module with:
```
module gitlab.com/amfiremage/lfstest.git
```

and 
```
require gitlab.com/amfiremage/lfstest.git-lfs v0.0.1
```
requires a module declare
```
module gitlab.com/amfiremage/lfstest.git-lfs 
```

**Comment by JAicewizard** [comment]
Ah that is good, as the original proposal/issue wasn't updated it seemed the original proposal was accepted. (I'm used to the original proposal being updated to keep up-to-date with the latest developments, or I've seen that happen in other proposals).

In that case I think I will put it in gerit when I have time

**Comment by lmb** [comment]
I'm currently looking at implementing this. The proposal doesn't specify whether `GOVCS` should include the proposed `git-lfs` by default. Any guidance on this?

**Comment by JAicewizard** [comment]
This is implemented. See https://github.com/JAicewizard/go/tree/lfs and https://github.com/golang/go/issues/47308#issuecomment-3694994875

**Comment by JAicewizard** [comment]
https://go-review.googlesource.com/c/go/+/733080/1 is the patchset currently active and rebased on master
